roadmap/state-sync-protocol.md

```diff
@@ -0,0 +1,198 @@
+
+Such a protocol should require a minimal amount of additional work from Execution nodes, and in particular it should not involve them doing a fan-out of the state they generate. Instead, Execution nodes should only need to answer a bounded number of initial state requests, after which the further exchange of this data should be delegated among Access nodes themselves. This allows us to leverage the scability of the Access node layer and gain increased data dissemination robustness and decentralization.
+
+## Design
+
+### Data Transfer
+
+The core data exchange protocol we build on top of is called [Bitswap](https://docs.ipfs.io/concepts/bitswap/). At a high level, Bitswap's main responsibilities are to acquire chunks of data requested by the client from the network, and send chunks of data in its possession to other peers who want them.
+
+Bitswap functions with a few fundamental concepts:
+* A *block* represents a raw chunk of data.
+* Each block is uniquely identified by a [*content identifier (CID)*](https://docs.ipfs.io/concepts/content-addressing/), which is based on a cryptographic hash of the block's contents.
+* A *blockstore* is a data store for storing and retrieving blocks.
+* A *session* orchestrates the flow of messages with other peers for discovering and retrieving the individual blocks which comprise a single high-level object.
+
+The full details of the Bitswap protocol can be found [here](https://research.protocol.ai/publications/accelerating-content-routing-with-bitswap-a-multi-path-file-transfer-protocol-in-ipfs-and-filecoin/delarocha2021.pdf).
+
+> To avoid conflicting terminology, we will henceforth refer to blocks in Bitswap as *blobs*.
+
+### State Computation
+
+Execution nodes, upon executing the transactions in each block, will compute the following data:
+
+```protobuf
+message ExecutionData {
+    repeated Collection collections = 1;
+    repeated TransactionResult txResults = 2;
+    repeated Event events = 3;
+    repeated TrieUpdate trieUpdates = 4;
+}
+```
+
+They will then serialize each of these fields separately<sup>[1](#footnotes)</sup> and compute a root CID from the serialized data for each field as follows:
+
+1. Split the data into blobs of up to 1MB (the message size limit of the [underlying networking stack](https://docs.ipfs.io/concepts/libp2p/)).
+2. Compute the CID of each blob and store it in the blobstore.
+3. If the number of CIDs computed in the previous step is 1, then return this CID. Otherwise, concatenate the CIDs and repeat from step 1<sup>[2](#footnotes)</sup>.
+
+**[INSERT DIAGRAM HERE]**
+
+At the end of this process, we are left with four root CIDs (one for each field in the Execution Data). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.
```

Comment:
We serialize each field and make them as a CID, so in the end there will be 4 CIDs, since there are 4 fields. But if a field is bigger than 1MB, which exceeds the networking limit, we will further split it into smaller CIDs. So in the end, we could have more than 4 CIDs. right?

In this case, when I receive 5 CIDs in total, how do I know which 2 CIDs is more the field that is over 1MB and need to be combined?


```suggestion
At the end of this process, we are left with four root CIDs (one for each field in the Execution Data, or more if one field is bigger than 1MB). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.
```

Reply:
Also, if a field is bigger than 1MB, how does the execution node let the access node know that there are 5 CIDs rather than 4 in total to be downloaded?

Reply:
@zhangchiqing 

No, we will only have a single CID, because the CID's are arranged in a tree. 

I added a diagram in the latest version, does that help clarify things?

I would also love input on how to improve the diagram :)